shell/server/drivers/mail.js

```diff
@@ -223,15 +240,23 @@ hackSendEmail = (session, email) => {
       };
     }
 
-    if (email.from.address !== grainAddress && email.from.address !== userAddress.address) {
+    if (removeValidPlusSuffix(email.from.address) !== grainAddress && email.from.address !== userAddress.address) {
       throw new Error(
-        "FROM header in outgoing emails need to equal either " + grainAddress + " or " +
+        "FROM header in outgoing emails need to equal either " + grainAddress + " (with optional suffix) or " +
         userAddress.address + ". Yours was: " + email.from.address);
     }
 
     // Unpack fields
     const { from, to, cc, bcc, replyTo, subject, text, html } = email;
 
+    // We allow envelope address to have a suffix as well. In this way apps can use unique envelope
+    // addresses when sending out e-mails which allow better handling of e-mail bounces - app can
+    // match bounce with the sent e-mail.
+    let envelopeFrom = email.envelopeFrom || from.address;
```

Comment:
The restriction on `From` is different since we actually allow the app to specify the user's e-mail address (rather than the grain's) as `From`. But it fundamentally does not make sense to allow overriding the envelope sender address to anything other than the grain's address, because the whole purpose of the envelope sender address is to point back at the sending entity.

Incidentally, depending on the e-mail provider in use, this may be moot. Sendgrid, for example, does not send bounces back to the sender; it adds the recipient address to the bounce list and silently drops all future messages to that recipient. So, on Oasis, your app will never receive bounces.

Generally I would like to shield apps from having to understand the complexity of e-mail processing. As you point out, Sandstorm already verifies addresses, so bounces will be uncommon. But if they do happen, that's a problem for Sandstorm, independent of the app. That is to say, regardless of what any app does, Sandstorm is going to need some sort of mechanism of its own for dealing with broken e-mail addresses. I'd prefer if apps can rely on Sandstorm taking care of these things and not try to do it themselves. (However, Sandstorm doesn't take care of these things today, so... this is admittedly vaporware.)

On another note, the notifications API still needs a lot of work, but our intent for that API is that it should allow "github-like" email notifications without the app having to understand any specifics of e-mail. See: https://github.com/sandstorm-io/sandstorm/tree/master/roadmap/platform/activity